File: 442.Find_All_Duplicates_in_an_Array/find-all-duplicates-in-an-array.py
# Source : https://leetcode.com/problems/find-all-duplicates-in-an-array/
# Author : Kyon Huang
# Date   : 2017-09-19

class Solution(object):
    def findDuplicates(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        res = []
        for x in nums:
            if nums[abs(x)-1] < 0:
                res.append(abs(x))
            else:
                nums[abs(x)-1] *= -1
        return res

File: 442.Find_All_Duplicates_in_an_Array /find-all-duplicates-in-an-array.java
// Source : https://leetcode.com/problems/find-all-duplicates-in-an-array/
// Author : Kyon Huang
// Date   : 2017-09-19

// 第一思路，非最优解法

class Solution {
    public List<Integer> findDuplicates(int[] nums) {
        List list = new ArrayList<Integer>();
        int len = nums.length;
        int[] arr = new int[len+1];
        for(int i = 0; i < len; i++) {
            arr[nums[i]]++;
        }
        for(int i = 1; i <= len; i++) {
            if(arr[i] > 1) {
                list.add(i);
            }
        }
        return list;
    }
}